### ISSUE TITLE
Unable to determine object type

### ISSUE THREAD

**Issue by strangerkir** [vimeo/psalm#7237]
It seems like Psalm is unable to determine the object type under some conditions.
Is this a bug or did I misunderstand something?

Simplified example: https://psalm.dev/r/3fca8ece29

**Comment by orklah** [comment]
Psalm does not like working with undefined variables, so doing this fixes the issue: https://psalm.dev/r/44e16a681e

However, I'm surprised I don't see a PossiblyUndefinedVariable in there, so this seems like a bug anyway

**Comment by weirdan** [comment]
>  I'm surprised I don't see a PossiblyUndefinedVariable in there

I'm pretty sure `??` suppresses it. And some will say it should.

**Comment by orklah** [comment]
Yeah, it's either a bug with ?? not being able to use the possibly undefined type for certain or a bug with Psalm not emitting the issue

**Comment by orklah** [comment]
Reduced to https://psalm.dev/r/10674e1851

Basically, at the end of a conditional. Psalm collect variables that changed or that were added in every branch of the conditional. However, in the snippet above, Psalm sucessfully detect that there is a possibly undefined variable that were created but it doesn't add it's type into vars_in_scope at all.

**Comment by orklah** [comment]
Ok, so this is harder than I thought.

Psalm has two ways of detecting possibly undefined variables:
- having a possibly_undefined flag on the Union type for that variable
- not knowing that variable at all

And it doesn't distinguish a lot between the two.

I tried to make it so the conditional registers the type with the flag instead of not registering it but it makes a lot of things fail. Then I tried to still emit the PossiblyUndefined Issue on CoalesceAnalyzer, but it's emitted in VariableFetchAnalyzer, in a place that believes we're in an isset() operator (Psalm simulate that), so changing that would create a lot of waves.

The thing we could do to mitigate that would be to check that the variable is known when we're inside the CoalesceAnalyzer, and if not, we could emit an Issue saying "Variable is possibly undefined, type was not inferred, please define the variable". But it's not very elegant...
